controllers/secret_controller.go

```diff
@@ -12,136 +12,117 @@ import (
+			return isUserDataSecret(e.Object)
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return isUserDataSecret(e.Object)
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			// get update event only when secret data is changed
+			if isUserDataSecret(e.ObjectNew) {
+				if string(e.ObjectOld.(*core.Secret).Data["userData"][:]) != string(e.ObjectNew.(*core.Secret).Data["userData"][:]) {
+					return true
+				}
+			}
+			return false
+		},
+	})
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&core.Secret{}, privateKeyPredicate).
+		Watches(&source.Kind{Type: &core.Secret{}}, handler.EnqueueRequestsFromMapFunc(r.mapToPrivateKeySecret), mappingPredicate).
+		Complete(r)
 }
 
-// isUserDataSecret returns true if the object meta indicates that the object is the userData Secret
-func isUserDataSecret(meta meta.Object) bool {
-	return meta.GetName() == userDataSecret && meta.GetNamespace() == userDataNamespace
+// isUserDataSecret returns true if the provided object is the userData Secret
+func isUserDataSecret(obj client.Object) bool {
+	return obj.GetName() == userDataSecret && obj.GetNamespace() == userDataNamespace
 }
 
-// isPrivateKeySecret returns true if the object meta indicates that the object is the private key secret
-func isPrivateKeySecret(meta meta.Object, keyNamespace string) bool {
-	return meta.GetName() == secrets.PrivateKeySecret && meta.GetNamespace() == keyNamespace
+// isPrivateKeySecret returns true if the provided object is the private key secret
+func isPrivateKeySecret(obj client.Object, keyNamespace string) bool {
+	return obj.GetName() == secrets.PrivateKeySecret && obj.GetNamespace() == keyNamespace
 }
 
-// blank assignment to verify that ReconcileSecret implements reconcile.Reconciler
-var _ reconcile.Reconciler = &ReconcileSecret{}
-
-// ReconcileSecret reconciles a Secret object
-type ReconcileSecret struct {
+// SecretReconciler is used to create a controller which manages Secret objects
+type SecretReconciler struct {
 	// This client, initialized using mgr.Client() above, is a split client
 	// that reads objects from the cache and writes to the apiserver
 	client client.Client
 	scheme *runtime.Scheme
 	log    logr.Logger
+	// recorder to generate events
+	recorder record.EventRecorder
+	// watchNamespace is the namespace the operator is watching as defined by the operator CSV
+	watchNamespace string
 }
 
 // Reconcile reads that state of the cluster for a Secret object and makes changes based on the state read
 // and what is in the Secret.Spec
 // The Controller will requeue the Request to be processed again if the returned error is non-nil or
 // Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
-func (r *ReconcileSecret) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+func (r *SecretReconciler) Reconcile(ctx context.Context, request ctrl.Request) (reconcile.Result, error) {
 	log := r.log.WithValues("secret", request.NamespacedName)
```

Comment:
is this required now? this has been initialized in NewSecretReconciler()

Reply:
Yep, this is adding the context of the secret being reconciled to a new logger instance